### DISCUSSION TITLE
Dependency Injection in Process Framework

### DISCUSSION THREAD

**Discussion by abhijeetagnihotri** [microsoft/semantic-kernel#9379] [General]
Hi Team!

I am experimenting for our future product using semantic kernel process flow framework. However, I don't find a good example to inject dependencies into the process steps in process framework. Please provide an example to use the same.

For reference, Below is the structure of my code.
IKernelBuilder kernelBuilder = Kernel.CreateBuilder();

kernelBuilder.AddAzureOpenAIChatCompletion("gpt-4o", "https://<custom>.openai.azure.com", "secret");
kernelBuilder.AddAzureOpenAITextEmbeddingGeneration("text-embedding-3-large", "https://<custom>.openai.azure.com/", "secret");
kernelBuilder.Services.AddScoped(m =>
{
    var embeddingService = m.GetRequiredService<ITextEmbeddingGenerationService>();
    return new AzureOpenAiEmbeddingService(embeddingService);
});

Kernel kernel = kernelBuilder.Build();

ProcessBuilder processBuilder = new("CustomProcess");

//Add steps
var embeddingStep = processBuilder.AddStepFromType<GetEmbeddingsFromTextStep>();

GetEmbeddingsFromTextStep.cs file has a parameterized constructor which expects a dependency to be injected.

**Comment by listentorick** [comment]
the kernal has GetRequiredService so, you might be able to use that

**Comment by alliscode** [comment]
Hi @abhijeetagnihotri, thanks for your question. Dependency injection works in the process steps as long as you have the Kernel configured with your DI, see here for an example: https://github.com/microsoft/semantic-kernel/blob/2fdd74e18d765260da663273708bd38134450ff1/dotnet/samples/GettingStarted/Step4_Dependency_Injection.cs#L13

Additionally, if you want to configure an initial state for a stateful step, you can do that like this:

```C#
var myCStep = processBuilder.AddStepFromType<CStep, CStepState>(initialState: new() { CurrentCycle = 1 });
``` 

See here: https://github.com/microsoft/semantic-kernel/blob/2fdd74e18d765260da663273708bd38134450ff1/dotnet/samples/Demos/ProcessWithDapr/Controllers/ProcessController.cs#L55
